python/tvm/contrib/graph_runtime.py

```diff
@@ -33,17 +33,45 @@ def create(graph_json_str, libmod, ctx):
             graph_json_str = graph_json_str._tvm_graph_json()
         except AttributeError:
             raise ValueError("Type %s is not supported" % type(graph_json_str))
-    device_type = ctx.device_type
-    device_id = ctx.device_id
-    if device_type >= rpc_base.RPC_SESS_MASK:
+    if isinstance(ctx, TVMContext):
+        ctx = [ctx]
+    elif not isinstance(ctx, (list, tuple)):
+        raise ValueError("ctx has to be the type of TVMContext or a list of "
+                         "TVMCTVMContext")
+    cpu_ctx_index = -1
+    for i, cur_ctx in enumerate(ctx):
+        if not isinstance(cur_ctx, TVMContext):
+            raise ValueError("ctx has to be the type of TVMContext or a list "
+                             "of TVMCTVMContext")
+        if cur_ctx.device_type == tvm.cpu(0).device_type:
+            cpu_ctx_index = i
+        elif cur_ctx.device_type >= rpc_base.RPC_SESS_MASK:
+            ctx[0], ctx[i] = ctx[i], ctx[0]
+
+    num_devices = len(ctx)
+    device_type_id = []
+    for cur_ctx in ctx:
+        device_type_id.append(cur_ctx.device_type)
+        device_type_id.append(cur_ctx.device_id)
+
+    if device_type_id[0] >= rpc_base.RPC_SESS_MASK:
         assert libmod.type_key == "rpc"
-        assert rpc_base._SessTableIndex(libmod) == ctx._rpc_sess._tbl_index
+        assert rpc_base._SessTableIndex(libmod) == ctx[0]._rpc_sess._tbl_index
         hmod = rpc_base._ModuleHandle(libmod)
-        fcreate = ctx._rpc_sess.get_function("tvm.graph_runtime.remote_create")
-        device_type = device_type % rpc_base.RPC_SESS_MASK
-        return GraphModule(fcreate(graph_json_str, hmod, device_type, device_id), ctx)
+        fcreate = ctx[0]._rpc_sess.get_function("tvm.graph_runtime.remote_create")
+        device_type_id[0] = device_type_id[0] % rpc_base.RPC_SESS_MASK
+        return GraphModule(fcreate(graph_json_str, hmod, num_devices,
+                                   *device_type_id), ctx[0])
+
+    # Assume CPU is the host processor when there are multiple devices on
+    # a hardware platform.
+    if (num_devices > 1) and (cpu_ctx_index < 0):
+        raise RuntimeError(
+            "CPU should be the host processor for heterogenous execution, but"
+            " not found in ctx.")
     fcreate = get_global_func("tvm.graph_runtime.create")
-    return GraphModule(fcreate(graph_json_str, libmod, device_type, device_id), ctx)
+    return GraphModule(fcreate(graph_json_str, libmod, num_devices,
```

Comment:
just pas in ctx[0]

Reply:
@tqchen So we actually pass fcreate(json, libmod, device_type_id[0], device_type_id[1], *device_type_id_others), right?

Reply:
yah, that can be a solution